src/wp-includes/media.php

```diff
@@ -5614,3 +5832,45 @@ function wp_increase_content_media_count( $amount = 1 ) {
 
 	return $content_media_count;
 }
+
+/**
+ * Add fetchpriority='high' to $loading_attrs if applicable.
+ *
+ * @since 6.3.0
+ * @access private
+ *
+ * @param array[] $loading_attrs Array of the loading attributes for the element.
+ * @param array[] $attr Array of the attributes for the element.
+ * @return array[] $loading_attrs Updated loading attributes for the element.
+ */
+function wp_maybe_add_fetchpriority_high_attr( $loading_attrs, $attr ) {
+	if ( ! empty( $attr['fetchpriority'] ) ) {
+		$loading_attrs['fetchpriority'] = $attr['fetchpriority'];
+		return $loading_attrs;
+	}
+	$img_size                   = $attr['width'] * $attr['height'];
+	$wp_min_priority_img_pixels = apply_filters( 'wp_min_priority_img_pixels', 5000 );
+	if ( $wp_min_priority_img_pixels <= $img_size && wp_high_priority_element_flag() ) {
+		$loading_attrs['fetchpriority'] = 'high';
+		wp_high_priority_element_flag( false );
+	}
+	return $loading_attrs;
+}
+
+/**
+ * A flag that indicates if a media is a possible candidate for fetchpriority='high'.
+ *
+ * @since 6.3.0
+ * @access private
+ *
+ * @param bool $value Optional. Used to change the static variable. Default null.
+ * @return bool $high_priority_element Return true if element is of higer priority else false.
+ */
+function wp_high_priority_element_flag( $value = null ) {
```

Comment:
I don't think we need an additional function for this. All of this logic can be incapsulated in `wp_maybe_add_fetchpriority_high_attr`.

Reply:
See https://github.com/WordPress/wordpress-develop/pull/4495#discussion_r1206778181, the reason this is needed is for tests to reset the flag.

Reply:
Appears I missed this thread during earlier reviews.

In terms of the tests, is it that one test was causing another to fail, or was it a failure during a single test?

Reply:
@joemcgill issue we were facing is once (trying to explain based on your suggested code) the static variable `$images_applied` once incremented can never be reset back to `0`. This is not a problem for production code but when it comes to multiple testcases we want to reset it back to `0` before every test, which is only possible if we have reset logic in place. @costdev even though reset logic is a overhead and is never needed in production I couldn't find a way to reset the static variable for testing.

Reply:
@joemcgill I acknowledge that the way this is handled here isn't great, at the same time I think your proposal would only move the problem somewhere else. Introducing a filter for something that is just relevant for tests is IMO not better than a private function, particularly because as far as I'm aware there isn't even a way to mark a filter as private, while with a function we can do that.

Also the way the function works is more specific to what the logic needs, which is just to set a flag when an image was marked with `fetchpriority="high"`. We don't need a count for that, I think that makes the logic actually harder to understand.

FWIW the existing `wp_increase_content_media_count()` function in core is a similarly specific function just to modify a static variable, so I think it's okay that we keep this for now. The alternative for both of them would be to just go with the classical `global` variable approach that is historically so common, yet somewhat discouraged in core. I think we can always change how this works in the future, and I'd argue in that case we should align how we do it for both `wp_increase_content_media_count()` and `wp_high_priority_element_flag()`.

Reply:
I think we should commit and fix after commit.

Reply:
@kt-12 If the static variable set by one test was causing issues when the next test ran as it couldn't be reset, then the `@runInSeparateProcess` and `@preserveGlobalState disabled` annotations can be added to tests that use the functions containing `static` variables.

The downside is that this slows the tests down as new processes are spun up. However, the alternative in the current timeframe is introducing a general method solely for the purpose of testing, which is worse than a few slow tests IMO.

With more time, abstracting a lot of the logic to classes is likely best, but with limited time, I'd suggest removing the "for tests only" general function and using the above annotations as a "for now" solution.